tests/test_dataclasses.py

```diff
@@ -801,6 +801,46 @@ class Config:
         e.item.name = 'pika2'
 
 
+def test_pydantic_callable_field():
+    """pydantic callable fields behaviour should be the same as stdlib dataclass"""
+
+    def foo(arg1, arg2):
+        return arg1, arg2
+
+    @pydantic.dataclasses.dataclass
+    class HasCallablesDC:
+        non_default_callable: Callable
+        default_callable: Callable = lambda x: foo(x, 'default')
+        default_callable_factory: Callable = dataclasses.field(default=lambda x: foo(x, 'factory'))
+
+    class HasCallablesModel(BaseModel):
+        non_default_callable: Callable
+        default_callable: Callable = lambda x: foo(x, 'default')
+        default_callable_factory: Callable = Field(default_factory=lambda: lambda x: foo(x, 'factory'))
+
+    @dataclasses.dataclass
+    class HasCallablesStdlibDC:
+        non_default_callable: Callable
+        default_callable: Callable = lambda x: foo(x, 'default')
+        default_callable_factory: Callable = dataclasses.field(default_factory=lambda: lambda x: foo(x, 'factory'))
+
+    def non_default_callable(x):
+        return foo(x, 'nondefault')
+
+    for cls in (HasCallablesModel, HasCallablesDC):
```

Comment:
it's really confusing if something fails to work out which object failed.

Better just to split the test and duplicate code, or if there's a lot of duplication, use fixtures.

Reply:
I copied the test from the issue but I agree it was a bit confusing. I rewrote the test. Hope it's easier to read.
Also added with an annotated callable like `Callable[[int, float, str], bool]`.
Ready for another review :)